clients/adapter/src/main/java/org/eclipse/hono/client/upstream/CommandResponse.java

```diff
@@ -0,0 +1,195 @@
+/*******************************************************************************
+ * Copyright (c) 2018, 2020 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+package org.eclipse.hono.client.upstream;
```

Comment:
Sticking to the downstream/upstream structure, I think this class as well as `CommandResponseSender` should be in the `downstream` package.

Reply:
I see your point. However, my idea was not so much to organize by technical aspects but rather by means of communication patterns. Most devices will support the telemetry (downstream) direction, i.e. they will report sensor readings etc. Some will additionally support receiving commands (upstream) and (optionally) responding to commands.
Maybe it would be better to rename the packages to `org.eclipse.hono.client.telemetry` and `org.eclipse.hono.client.command` respectively? WDYT @calohmn

Reply:
I think I would still prefer the downstream/upstream packages, with "CommandResponseSender" in the downstream package.
To me, grouping classes by message flow direction offers more value here than using the communication pattern as the grouping theme. For the request/response pattern (#1276) we might also reuse the existing senders/consumers, so the Hono communication patterns wouldn't be directly reflected in the packages here anyway.
And having `org.eclipse.hono.client.telemetry` contain telemetry *and* event senders is also a bit odd. A different term encompassing both telemetry and events would be better, although I guess we don't have that.

Reply:
> And having org.eclipse.hono.client.telemetry contain telemetry and event senders is also a bit odd. A different term encompassing both telemetry and events would be better, although I guess we don't have that.

We have been calling it the *telemetry* direction from the very beginning ...

> For the request/response pattern (#1276) we might also reuse the existing senders/consumers, so the Hono communication patterns wouldn't be directly reflected in the packages here anyway.

Well, FMPOV re-designing the client provides us with the opportunity to provide interfaces that are taylored to the the use case at hand. If we want to support a request/response pattern then FMPOV we should add a new package with new interfaces that contain method names and wording describing the actual problem being addressed. This should make it much easier to understand the exchange pattern compared to re-using the Telemetry and Command related interfaces for this purpose.

Reply:
@calohmn can we agree to group the interfaces into packages based on semantics instead of (technical) message flow direction? And then rename packages *downstream* and *upstream* to *telemetry* and *command* respectively for the time being and see if we find better names in the future?

Reply:
Yes, I'm ok with using *telemetry* and *command* for now.